ocaml/xapi/xapi_guest_agent.ml

```diff
@@ -112,10 +124,56 @@ let networks path (list: string -> string list) =
          | Some pair -> pair :: acc
       ) [] (list path)
   in
-  path
-  |> find_eths
-  |> List.map (fun (path, prefix) -> find_all_ips path prefix)
-  |> List.concat
+  let find_vifs vif_path =
+    let extract_vif acc vif_id = ((extend vif_path vif_id), vif_id) :: acc in  
+    List.fold_left extract_vif [] (list vif_path)
+  in
+  let compare a b = 
+    let open Pervasives in
+    compare (int_of_string a) (int_of_string b) 
+  in
+  let find_all_vif_ips vif_path vif_id = 
+    (*  vif_path: attr/vif/0 *)
+    (*  vif_id: 0 *)
+    let extract_ip_ver vif_id acc ip_ver = 
+      let ip_addr_ids = list (extend vif_path ip_ver)  in
+      let extract_ip_addr vif_id ip_ver acc ip_addr_id = 
+        let key_left = Printf.sprintf "%s/%s/%s" vif_path ip_ver ip_addr_id in
+        let key_right = Printf.sprintf "%s/%s/%s" vif_id ip_ver ip_addr_id in
+        match acc with
+        | [] when ip_ver = "ipv4"  -> 
+          [(key_left, (extend vif_id "ip")); (key_left, key_right)]
+        | _ -> (key_left, key_right) :: acc
+      in  
+      try 
+        List.fold_left (extract_ip_addr vif_id ip_ver) 
+          acc (List.stable_sort compare ip_addr_ids)
+      with Failure "int_of_string" ->
+        error "Invalid vif ID.";
+        []
+    in
+    let ip_vers = List.filter (fun a -> a = "ipv4" || a = "ipv6") (list vif_path) in
+    List.fold_left (extract_ip_ver vif_id) [] ip_vers
+  in
+  let iface_pair_list = find_eths path in
+  let vif_pair_list = find_vifs (extend path "vif") in
+  let has_iface = (List.length iface_pair_list) > 0 in
```

Comment:
Computing the `length` is not efficient. You could define `let not_empty = function [] -> false | _ -> true` and use `not_empty iface_pair_list`.